src/main/java/edu/umich/tl/analytics/CanvasCourseReportServlet.java

```diff
@@ -74,32 +84,121 @@ protected void doGet(HttpServletRequest request,HttpServletResponse response){
+			response.setHeader("Content-Disposition","attachment;filename=PublishedCanvasCoursesWithInstructorsFor"+courseReportTask.termName+".csv");
+			String outputResult = generateCSVFile(cfs).toString();
+			courseReportReqThreads.remove(aCourseReportThread);
+			outputStream.write(outputResult.getBytes());
+			outputStream.flush();
+			outputStream.close();
+			M_log.debug("Ending polling... Thread size: "+courseReportReqThreads.size());
+		}
+	}
 	private void getPublishedCourses(HttpServletRequest request,HttpServletResponse response) {
-		M_log.debug("getPublishedCourses: Called");
+		M_log.debug("getPublishedCourses(): Called");
 		String term = request.getParameter(TERM);
 		String termName = request.getParameter(TERM_NAME);
-		String subAccountUrl = canvasURL+"/api/v1/accounts/1/sub_accounts?recursive=true&per_page=100";
-		String url = canvasURL+"/api/v1/accounts/1/courses?enrollment_term_id="+term+"&published=true&per_page=100";
 		CoursesForSubaccounts cfs =new CoursesForSubaccounts();
-		getSubAccountInfo(subAccountUrl,cfs,response);
-		if(!cfs.subaccountCallHasErr) {
-			getThePublishedCourseList(url,cfs,response);
-			if(!cfs.courseCallHasErr) {
-				response.setContentType("text/csv");
-				response.setHeader("Content-Disposition","attachment;filename=PublishedCanvasCoursesWithInstructorsFor"+termName+".csv");
-				try {
-					OutputStream outputStream = response.getOutputStream();
-					String outputResult = generateCSVFile(cfs).toString();
-					outputStream.write(outputResult.getBytes());
-					outputStream.flush();
-					outputStream.close();
-				} catch (IOException e) {
-					M_log.error("Writing the csv file to the has some errror: ",e);
-				}
-			}
+		CourseReportTask courseReportTask=new CourseReportTask(term, termName,cfs);
+		Thread courseReportThread=new Thread(courseReportTask);
+		courseReportReqThreads.put(courseReportThread, courseReportTask);
+		courseReportThread.start();
+		PrintWriter out;
+		try {
+			out = response.getWriter();
+			out.print(courseReportThread.getId());
+			out.flush();
+		} catch (IOException e) {
+			e.printStackTrace();
 		}
 	}
 
+	public class CourseReportTask implements Runnable{
+		private String term;
+		private String termName;
+		private CoursesForSubaccounts cfs;
+		
+
+		public CourseReportTask(String term, String termName,CoursesForSubaccounts cfs) {
+			this.term=term;
+			this.termName=termName;
+			this.cfs=cfs;
+		}
+		
+		public CoursesForSubaccounts getCfs() {
+			return cfs;
+		}
+
+		public void setCfs(CoursesForSubaccounts cfs) {
+			this.cfs = cfs;
+		}
+
+		public void run() {
+			getPublishedCourseThreadCall(term, termName,cfs);
+		}
+		
+	}
+	private CoursesForSubaccounts getPublishedCourseThreadCall(String term, String termName,CoursesForSubaccounts cfs) {
+		String subAccountUrl = canvasURL+"/api/v1/accounts/1/sub_accounts?recursive=true&per_page=100";
+		String url = canvasURL+"/api/v1/accounts/1/courses?enrollment_term_id="+term+"&published=true&per_page=100";
```

Comment:
IMHO, the current setup is ok for this release.  For the future consider making an 'URLCreator' class so which can take in the required parameters for a query and return full URLS when required.  This would:
- allow putting common strings in properties or, at least, all in one place,
- makes the intent of the URL clearer since nobody has to parse through the URL to see exactly what it's doing,
- put all the URL generation in the same place so it is easier to find and maintain URLs,
- make URL generation easy to unit test.  This gets more and more important the more requests we have to Canvas, or other external services.